kfp_tests/run_integration_tests.py

```diff
@@ -33,46 +34,60 @@
+        return -1
+    start = output.find("run_id|") + len("run_id|")
+    end = output.find("|end_id")
+    run_id = output[start:end]
+    return run_id
 
 
 def obtain_flow_file_paths(flow_dir_path):
     file_paths = [f for f in listdir(flow_dir_path) if isfile(join(flow_dir_path, f))]
-    file_paths.remove("__init__.py")  # __init__.py needed to establish Python module
     return file_paths
 
 
+@ray.remote
+def spawn_and_return(args, flow_file_path):
+    """
+    This function is the creation and waiting one KFP run. Since
+    we don't want to wait for the entire flow to finish on KFP before
+    creating a new run, we will have Ray parallelize the effort.
+    """
+    full_path = join(args.flow_dir_path, flow_file_path)
+    process = run(
+        ["python", full_path, "--datastore=s3", "kfp", "run"],
+        stdout=PIPE,
+        stderr=PIPE,
+        check=True,
+        text=True,
+    )
+    run_id = parse_run_id(process.stdout)
+    if run_id != -1:
+        run_link = f"{KFP_RUN_URL_PREFIX}/_/pipeline/#/runs/details/{run_id}"
+        print(f"Running {flow_file_path} ... Run link: {run_link}")
+        kfp_client = kfp.Client(namespace=args.namespace, userid=args.userid)
+        run_response = kfp_client.wait_for_run_completion(run_id, 500).to_dict()
+        if run_response["run"]["status"] == "Succeeded":
+            print(f"Flow {flow_file_path} ran successfully! Run link: {run_link}")
+        else:
+            print(
+                f"Flow {flow_file_path} encountered a runtime error. Run link: {run_link}"
+            )
+    else:
+        print(f"Flow {flow_file_path} failed to compile properly.")
+    return
+
+
 def test_sample_flows(args):
     flow_file_paths = obtain_flow_file_paths(args.flow_dir_path)
-    kfp_client = kfp.Client(namespace=args.namespace, userid=args.userid)
-    for flow_file_path in flow_file_paths:
-        full_path = join(args.flow_dir_path, flow_file_path)
-        print(f"Running {flow_file_path} ...")
-        process = run(
-            ["python", full_path, "--datastore=s3", "kfp", "run"],
-            stdout=PIPE,
-            stderr=PIPE,
-            check=True,
-            text=True,
-        )
-        run_id = parse_run_id(process.stdout)
-        if run_id != -1:
-            run_response = kfp_client.wait_for_run_completion(run_id, 500).to_dict()
-            if run_response["run"]["status"] == "Succeeded":
-                print(f"Flow {flow_file_path} ran successfully!")
-            else:
-                print(f"Flow {flow_file_path} encountered a runtime error.")
-        else:
-            print(f"Flow {flow_file_path} failed to compile properly.")
+    task_list = []
+    for (
+        flow_file_path
+    ) in flow_file_paths:  # we append tasks, which immediately return a "future"
+        task_list.append(spawn_and_return.remote(args, flow_file_path))
+    ray.wait(task_list, num_returns=len(task_list))  # wait for all tests to complete
+    return
```

Comment:
We may loop through results to see if any test failed. In most case this will help avoid manually checking for each pipeline run result.

Reply:
@cloudw can you explain what you mean by `loop through results to see if any test failed`. This comment might not be relevant since we don't use Ray anymore. But, even in PyTest, when the each KFP run is created, we check for the returncode of the spawning process (`python flow.py kfp run`) first. If the returncode is non-zero, the process exits and we don't wait for any run to complete.

Reply:
In the version of code commented, you waited for the results but didn't check for return values. This again relates to how exceptions are handled, since exception from subprocess is returned when you tries to get result of a future value. 
For pytest you probably will not worry about this issue.

Reply:
I see. In the PyTest implementation, I first check for the return value of the subprocess (which is either returned after the KFP run completes or an exception is encountered). In both cases, I ensure the return code is 0, and if the KFP run fails prematurely, the return code is provided fairly quickly. I believe this comment has been resolved.